### ISSUE TITLE
Unified Push support

### ISSUE THREAD

**Issue by BigmenPixel0** [mattermost/mattermost-mobile#6673]
What about this?
unifiedpush.org

**Comment by Mannshoch** [comment]
I already use it for a matrix client. It works !

**Comment by erinarbleizez** [comment]
As a LineageOS user in order to de-google my phone because that's the best thing to do in this dark time (and without Micro-G possibility because my phone is not supported yet) I would really like having my notification pushed through unifiedpush (which is the best simple alternative to firebase cloud messaging).

It would also be more in accordance with F-Droid philosophy of not depends on Google.

I know nothing in android dev and wouldn't be any help but if there is anything server side code related that can be made to make it happen, just tell me I'm more than willing to help.

**Comment by rudism** [comment]
Supporting unified push would have several benefits:
- Push notifications would become available on devices without Google Services.
- Client builds from 3rd parties (such as [F-Droid](https://f-droid.org)) would gain push notification capabilities.
- Team Edition self-hosters could also self-host push servers without needing to compile their own client apps.
- Opening up to other 3rd party unified push providers would reduce the burden/dependency on the official Mattermost TPNS.
- Unified push is already used and battle-tested by many other FOSS projects (clients for Matrix, Mastodon, NextCloud, and so on).

**Comment by toastal** [comment]
I self-host my own UnifiedPush server via XMPP to disconnect as much data as possible from Google and other untrustworthy parties&mdash;both [Prosody](https://prosody.im) and [ejabberd](https://www.ejabberd.im) make it trivial to enable for the server side, & all [Conversations](https://conversations.im)-derived (such as [Cheogram](https://cheogram.com)) Android clients support the client side. I’ve been using it with other apps from F-Droid just fine (Element, Tusky) & Mattermost would help with work-related chat.

**Comment by TheJonny** [comment]
It seems mattermost is using https://github.com/wix/react-native-notifications/ for push notifications - so unifiedpush should be integrated there

**Comment by TheJonny** [comment]
Hi :)

Version 9 of the server supports a plugin hook for push notifications, so i did a bit of botching:
 - the server pluging https://git.sr.ht/~thejonny/mattermost-plugin-unifiedpush creates a slash command `/register_unifiedpush [url]`
 - the android side is a separate app, https://git.sr.ht/~thejonny/matterpush , as I am not confident in react. It displays the `/register_unifiedpush`-command needed, and displays notifications. if you tap the notification, it opens the mattermost:// link to open the correct chat in the mattermost app.
  It would be really awesome if somebody would help me integrating this. I'm available as `@thejonny:bruckbu.de` on matrix.
  
essential missing features are:
 -  [ ] encryption of the push messages
 - [ ]  rate limiting on the server to prevent abusing it for denial of service amplification: with n messages pushed to m sessions, the server plugin fires n*m GET requests to arbitrary URLs.
  - [ ] cleaning up stale/failed endpoints
  - [ ] app side: notify the user, if the endpoint changes, so they can reregister

**Comment by toastal** [comment]
Excellent work @TheJonny I appreciate that you’re hosting the code on a libre code forge too with a decentralized chat option listed.

**Comment by tgly307** [comment]
@TheJonny Thank you very much for your contribution. Could you provide the packed file, please? This will be a great help.

**Comment by TheJonny** [comment]
@tgly307 please do not expect stability yet :)
For Android, it's in this F-Droid repo: https://platon.bruckbu.de/~jonny/x/fdroid/repo/
the server part should be `go get; make`.
I don't want to provide a package right now, it should only be used in controlled environments or for testing, as the security can still be substantially improved:
 - you cannot logout (after logging out from the Mattermost App, you still receive Notifications via UnifiedPush
 - by registering n sessions and sending m messages, the server makes  n*m HTTP-requests to user controlled URLs, so it could be used for DOS amplification.
both is easy to fix, but I did not find the time yet, but maybe next week

**Comment by TheJonny** [comment]
@toastal now I have - I am looking forward for your issues :)

why wouldn't they spy - as it would be a no-brainer for apple or google to end-to-end-encrypt the push messages in the librariy (just provide a key along the device identification token), I consider it intentional.

**Comment by NCPSNetworks** [comment]
Hey, I've noticed that there doesn't seem to be much progress. I have a suggestion: add a background service (a long connection) to receive push notifications, similar to Mattermost Desktop. This would consume more power, but the option would be disabled by default, and users could enable it to use in regions without GMS installed or where FCM is blocked. I think this should be relatively easy to implement (Apple devices don't have this issue, so support only needs to be added for the Android version).

**Comment by NCPSNetworks** [comment]
Although this might increase power consumption on mobile devices, I think it's manageable (since Mattermost Mobile is a lightweight app, and the network overhead can be minimal). 

It can serve as a fallback solution for devices or regions that don't support FCM/GMS and is easier to implement. 

Mobile devices would only need a long connection to the Mattermost server to bypass FCM.

If you think this idea is good, could you create a ticket for it? Thank you!

**Comment by mirko** [comment]
Any news by any chance on this?

**Comment by curiousercreative** [comment]
Came here to ask the same. Hoping to get all my push notifications through my xmpp connection.

**Comment by bilogic** [comment]
yea this would be great

**Comment by curiousercreative** [comment]
I've just a few days ago begun bridging my Mattermost instance to my Snikket/Prosody instance (via slidge matteridge) which does unified push via Conversations Android client (I'm actually using Monocles via F-droid).

**Comment by curiousercreative** [comment]
@bilogic not exactly. I've set up a server bridge between my XMPP server and my Mattermost server. Now, it's possible for any users that are registered on both my XMPP and Mattermost server to connect their Mattermost account. With that server-server connection, an XMPP client will gain the ability to message with the Mattermost users and channels the connected account has access to (public and private channels). See [features supported here](https://slidge.im/docs/matteridge/main/user/features.html) (Multi-User Chat appears to be implemented rather than planned). I'm still testing the bridge for myself, but I could consider disabling mobile notifications for Mattermost and instead rely on my XMPP client.

I don't know what polling and other Mattermost specific UI translates, but I have one custom bot that posts using advanced formatting and those messages don't seem to make it through to XMPP in tact. I do get notified when there's a message, but it doesn't show in XMPP chat.

**Comment by bilogic** [comment]
@curiousercreative thanks but such a setup is a little beyond my users.. I wished mattermost could support some self hosted push service

**Comment by distinctjuggle** [comment]
It would be really nice to have this functionality.  It's the only thing preventing me from swapping over to Mattermost.

**Comment by amyblais** [comment]
Thanks, appreciate your feedback @distinctjuggle and everyone.

Would you like to [contribute this in the feature idea forum](https://mattermost.uservoice.com/forums/306457-general/) so it can be discussed, upvoted and considered for a [help wanted ticket](https://docs.mattermost.com/process/help-wanted.html)?

Please include a link back to this GitHub issue. If you're interested in implementing, please say so and we'll prioritize the review. 

You get **10** votes in the feature idea forum, and each one influences the future of the project.

**Comment by woiwoiwoiwoi** [comment]
Unified Push has gained a fair bit of momentum lately. Nextcloud has just rolledout support for it, most Matrix clients have been supporting it for a long time, and there are many more.

Being able to not have Google listen in on Mattermost conversations would be really really nice.

**Comment by apixandru** [comment]
I experimented today and got a proof of concept working with ntfy.

The mattermost architecture is solid. The only change needed is to the mattermost-push-proxy. By plugging in a custom, self-hosted mattermost-publish-proxy into the official, unmodified self-hosted Mattermost service, you have all the server-side components you need.

This isn’t a simple "well if you have something running, just merge whatever you have" situation because the mechanism exposes credentials that must be handled carefully, including unified push credentials and web push credentials, among others.

The challenge is not just about getting everything to work in the optimistic scenario; you must also account for all the failure scenarios and how things can go wrong.

**Comment by TheJonny** [comment]
The app uses a [patched](https://github.com/mattermost/mattermost-mobile/blob/main/patches/react-native-notifications%2B5.2.2.patch) variant of  https://github.com/wix/react-native-notifications for notification support.  So for the rouch road map would be
 - [ ] lift the react-native-notifications patches to it's source code (at the moment, typescript generated javascript is patched)
 - [ ] implement unified push in this library https://github.com/wix/react-native-notifications/issues/1000 (don't get distracted from the purple checkmark... we could copy the issue to an open one)
 - [ ] integrate it into the app. This means implementing a setting to choose between google push and unified push.
 - [ ] implement the server side. this should be fun and easy

**Comment by TheJonny** [comment]
> This isn’t a simple "well if you have something running, just merge whatever you have" situation because the mechanism exposes credentials that must be handled carefully, including unified push credentials and web push credentials, among others.

I agree, but the current approach is "send the users' conversations to google/apple", so it can't get much worse. using a unifiedpush library gives you end to end encryption for free

**Comment by apixandru** [comment]
as you can see by the closed PR, they are currently not looking into unified push support

in case anyone is looking into this as a personal endeavor, here is my POC for it

https://github.com/apixandru/mattermost-mobile/tree/unified_push
https://github.com/apixandru/mattermost-push-proxy/tree/unified_push

you can use the official mattermost server with the custom push proxy and the modified mobile app to get it running.

this is obviously at the POC state so it's not production grade, more like a fun side-project at the moment in case someone else wants to build on top of it.

**Comment by woiwoiwoiwoi** [comment]
To be honest, I don't understand their decision at all. They can of course do whatever they want - I just don't understand it. This is the main headline of their website:

**"Operational Sovereignty for National Security and Critical Infrastructure"**

And then they not only rely on proprietary big tech infrastructure, but don't even allow for a **SOVEREIGN** alternative.
